src/Raven.Server/Smuggler/Documents/ShardedDatabaseSmuggler.cs

```diff
@@ -0,0 +1,700 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Raven.Client.Documents.Indexes;
+using Raven.Client.Documents.Operations;
+using Raven.Client.Documents.Smuggler;
+using Raven.Client.ServerWide;
+using Raven.Client.Util;
+using Raven.Server.Config;
+using Raven.Server.Documents;
+using Raven.Server.Documents.Indexes;
+using Raven.Server.Documents.Indexes.Auto;
+using Raven.Server.Documents.Indexes.MapReduce.Auto;
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Commands;
+using Raven.Server.ServerWide.Commands.Indexes;
+using Raven.Server.ServerWide.Context;
+using Raven.Server.Smuggler.Documents.Data;
+using Raven.Server.Smuggler.Documents.Processors;
+using Raven.Server.Utils;
+using Sparrow.Json;
+using Sparrow.Logging;
+
+namespace Raven.Server.Smuggler.Documents
+{
+    public class ShardedDatabaseSmuggler : BaseSmuggler 
+    {
+        private readonly TransactionOperationContext _tContext;
+        private readonly List<DatabaseRecord.ShardRangeAssignment> _shardAllocation;
+        private List<ISmugglerDestination> _destinations;
+        private readonly DatabaseRecord _databaseRecord;
+        private readonly ServerStore _server;
+        
+        public ShardedDatabaseSmuggler( ISmugglerSource source, 
+            List<ISmugglerDestination> destination, 
```

Comment:
```suggestion
            List<ISmugglerDestination> destinations, 
```

Reply:
are they in the order of the shard allocations? How do we assign responsibility to each of them?

Reply:
Same order. every shard handle only item relevant to him

Reply:
Same